pkg/networkservice/chains/client/client.go

```diff
@@ -74,19 +80,28 @@ func NewClient(ctx context.Context, name string, onHeal *networkservice.NetworkS
 	return rv
 }
 
-// NewCrossConnectClientFactory - returns a (2.) case func(cc grpc.ClientConnInterface) NSM client factory.
-func NewCrossConnectClientFactory(name string, onHeal *networkservice.NetworkServiceClient, tokenGenerator token.GeneratorFunc, additionalFunctionality ...networkservice.NetworkServiceClient) func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
+// Supplier is a type for networkservice.NetworkServiceClient supplier function
+type Supplier func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient
+
+// NewClientFactory - returns a (2.), (3.) cases func(cc grpc.ClientConnInterface) NSM client factory.
+func NewClientFactory(
+	name string,
+	onHeal *networkservice.NetworkServiceClient,
+	tokenGenerator token.GeneratorFunc,
+	additionalFunctionalitySuppliers ...Supplier,
+) func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
 	return func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
-		return chain.NewNetworkServiceClient(
-			mechanismtranslation.NewClient(),
-			NewClient(ctx, name, onHeal, tokenGenerator, cc, additionalFunctionality...),
-		)
+		var additionalFunctionality []networkservice.NetworkServiceClient
+		for _, supplier := range additionalFunctionalitySuppliers {
+			additionalFunctionality = append(additionalFunctionality, supplier(ctx, cc))
+		}
+		return NewClient(ctx, name, onHeal, tokenGenerator, cc, additionalFunctionality...)
 	}
 }
 
-// NewClientFactory - returns a (3.) case func(cc grpc.ClientConnInterface) NSM client factory.
-func NewClientFactory(name string, onHeal *networkservice.NetworkServiceClient, tokenGenerator token.GeneratorFunc, additionalFunctionality ...networkservice.NetworkServiceClient) func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
-	return func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
-		return NewClient(ctx, name, onHeal, tokenGenerator, cc, additionalFunctionality...)
+// FromConstructor is a common networkservice.NetworkServiceClient constructor wrapper to Supplier
+func FromConstructor(constructor func() networkservice.NetworkServiceClient) Supplier {
+	return func(_ context.Context, _ grpc.ClientConnInterface) networkservice.NetworkServiceClient {
+		return constructor()
```

Comment:
So why all this additional baggage when we can simply use the func(_ context.Context, _ grpc.ClientConnInterface) networkservice.NetworkServiceClient